CONTRIBUTING.md

```diff
@@ -36,6 +36,38 @@ For the moment it would not seem appropriate for our plugin to throw instances c
+
+Implementation wise, we do this by passing a reference to the `FlutterResult` we can use to pass back the result, when. This makes it convenient for users: they can `await push.activate()`. However, they should not rely on this Future completing, in the case of network issues.
+
+#### Notification tap handling
+
+Android's firebase-messaging/ FCM library allows users to select the Intent action used when the automatically generated notification is tapped by the user. You can do this by setting `fcm.notification.click_action` in Ably's push payload. However, for this to work, users would need to declare the intent action within their `AndroidManifest.xml`. Therefore, we don't really tell users they can modify `click_action` and configure it. However, they can do so if they wish.
+
+#### Notifications generation
+
+iOS allows users to show the notification even if the app is in the foreground, by calling a delegate method ([`userNotificationCenter(_:willPresent:withCompletionHandler:)`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter)) which the user can choose to show the message to the user, based on the notification content. FCM/ Android does not provide this functionality. To be consistent with iOS, Ably listens for messages with notifications, and asks the user on the dart side if they'd like this notification shown, and if so, show it on behalf of the user. Because of this, we have a large file containing code taken from Firebase messaging's Android SDK, stored within `NotificationUtilities.java`.
+
+#### Push Notifications Background Message Handling
+
+Background processing in Flutter is a complicated subject that has not been explained very much in public. It involves creating Flutter Engines/isolates manually (for Android), passing messages back and forth, etc.
+
+Differences between Ably-Flutter and Firebase Messaging implementation (Android only):
+- **Entrypoint:** In Firebase messaging, they explicitly define a new DartEntrypoint (e.g. A Callback which is launched in it's own isolate) which will be launched by the Dart VM. We do not launch our custom isolate / entrypoint, we re-use the users default entry point (their application). In Ably-Flutter, we use re-use the users main application. Therefore, we provide the same environment for message handling on both Android and iOS: their application **is running** when we handle the message.
```

Comment:
Actually, my language is incorrect. Callbacks != Entrypoints. We use the default entrypoint, where as Flutter uses a callback. They are for the same purpose: starting an isolate with specified dart code. Callback is a dart function, entrypoint is a top level main function.

Reply:
215976dcaffff7f9a2e07ff5d1d7439b939c5dd7